File: week13/15/1_Mon_stone_HeapQ.py
import heapq
from typing import List

class Solution:
    def lastStoneWeight(self, stones: List[int]) -> int:
        h = []
        for stone in stones:
            heapq.heappush(h, -stone)   # heapq는 minheap

        while len(h) > 1:
            s1 = heapq.heappop(h)
            s2 = heapq.heappop(h)
            if s1 < s2:
                heapq.heappush(h, s1-s2)
        
        if len(h) > 0:
            return -h[0]
        else:
            return 0

File: week13/15/1_Mon_stone_HeapQ.cpp
#include <queue>

using namespace std;

class Solution {
public:
    int lastStoneWeight(vector<int>& stones) {
        priority_queue <int> Q;                     // priority_queue 는 Max Heap
        for (int stone : stones) Q.push(stone);

        while (Q.size() > 1) {
            int s1 = Q.top();
            Q.pop();
            int s2 = Q.top();
            Q.pop();
            if (s1 > s2) Q.push(s1-s2);            
        }

        return (Q.size() > 0) ? Q.top() : 0;
    }
};